### DISCUSSION TITLE
Remove assets and asset-packagist.org

### DISCUSSION THREAD

**Discussion by mtangoo** [yiisoft/yii2#20164] [Yii 22]
As part of Yii 2.2, it is good to remove assets from the core and make it a separate package. 
This discussion is to get ideas on what is the best way to execute this task.

Feel free to share whatever you think will be the best approach.

**Comment by schmunk42** [comment]
How would you handle `AssetBundle`s? 
Literally every extension uses them.

**Reply by mtangoo** [reply]
Can you explain why you think it is going to be a problem?

**Reply by schmunk42** [reply]
If you remove assets from the core, who takes care about `AssetBundle` classes?
How are assets registered then?

**Reply by terabytesoftw** [reply]
If it is correct, if we move the assets to separate packages, it will break everything, because the dependencies must be added to the extensions.

Now this can give the advantage of using:

1.- Asset packageagist.
2.- Foxy.
3. Any JS download manager (`Npm`, `Yarn`, `Pnpm`, `Bun`).

In fact fork foxy, and i am maintaining it, we can do some customization.

On the other hand, the Bower packages must be changed to `NPM`, and `JQUERY4` will break the framework.

The question is the `2.2` branch is just a mirror to improve some things, or are we going to continue moving forward without breaking anything.

**Reply by schmunk42** [reply]
> or are we going to continue moving forward without breaking anything.

For me, that would be crucial, if Yii 2.2 is not (more or less) backward-compatible - why not use 3.0 directly?

**Reply by terabytesoftw** [reply]
Yii2 has around 300,000/500,000 monthly downloads, which are all non-CI, many companies and developers use Yii2 in new projects because it is a Full Stack based on PHP+Bootstrap+Jquery, and it is one of the few that works very well.

**Reply by mtangoo** [reply]
@schmunk42 am still not getting your statement
> If you remove assets from the core, who takes care about AssetBundle classes?
How are assets registered then?

As far as I understood the whole thing, we are going to extract assets into a separate package which we can then change the way it "internally works" like how it fetches js packages.

This might be simplistic explanation but I still fail to see how it will affect AssetBundles. Can you pick an example that explains your point?

**Reply by mtangoo** [reply]
>For me, that would be crucial, if Yii 2.2 is not (more or less) backward-compatible - why not use 3.0 directly?

Isn't Yii2.2 a bridge to Yii3? So breaking changes should be ones that moves you closer to Yii3. Only breaking changes should not be huge like migrating to Yii3.

**Reply by schmunk42** [reply]
> As far as I understood the whole thing, we are going to extract assets into a separate package which we can then change the way it "internally works" like how it fetches js packages.

How js-packages are fetched (or installed) is part of the composer installation process, there's nothing you can change in the framework about it besides removing `bower` or `npm` assets from `composer.json`. This has to be done in every package or with some `provides` hacks from "a plugin".

And `npm` has it's own glitches, eg. often packages *have to* use different versions of the same js-lib, which is not problem with npm or yarn, but impossible with `composer`.

See this post for more details https://github.com/yiisoft/assets/issues/7#issue-540430671

The first thing would be to switch to **foxy** while still having a way to deal with `@bower/..` and `@npm/...` packages, eg. they should not install anything, but you can somehow forward this task to foxy.

**Reply by mtangoo** [reply]
>How js-packages are fetched (or installed) is part of the composer installation process, there's nothing you can change in the framework about it besides removing bower or npm assets from composer.json. This has to be done in every package or with some provides hacks from "a plugin".

The whole point of 2.2 is allowing BC like this one. Yes composer JSON will have to be updated whether manually or with plugin. Never done composer plugins, but I guess it should not be complex for this one. Check Yii version, if 2.2.x then update json file accordingly

**Reply by schmunk42** [reply]
Just for the record: I just tried to install foxy in our base template (phd5, yii2) and it failed, same as https://github.com/fxpio/foxy/issues/52
I also could not get @terabytesoftw maintained fork (https://github.com/php-forge/foxy) to run.

The idea would have been to have foxy ready as a plugin and then successively replace `@bower-asset`, `@npm-asset` with `package.json`s.

@yiisoft/core-developers What's the actual status of client side asset installation in Yii3 btw.? Is this supposed to work?

**Reply by terabytesoftw** [reply]
It's strange i use https://github.com/php-forge/foxy in https://github.com/yii3-extensions/app and it works correctly even with bun, which was the error.

https://github.com/yii3-extensions/app/blob/8ec3ac6455be5fab228d796e82ffee4219e86a2f/composer.json#L115

**Reply by terabytesoftw** [reply]
btw in Yii3 there is no download manager for the assets, you use whatever you want, from `asset-packagist.org`, `foxy`, or simply `npm`.

**Reply by schmunk42** [reply]
> btw in Yii3 there is no download manager for the assets, you use whatever you want, from asset-packagist.org, foxy, or simply npm.

How do you use `asset-packagist.org` with Yii3, could you point me to the docs?


>  works correctly even with bun, which was the error.

First I tried `yarn`, and got:

```
Merging Composer dependencies in the asset package
Installing yarn dependencies
Usage: yarn [options]

yarn: error: no such option: --non-interactive
Fallback to previous state for the Asset package
Fallback to previous state for Composer

In Locker.php line 352:
                                 
  Undefined array key "version"  
```

which looks like an incompatible `yarn` version?!

Then I tried to move to `npm` and got: 

```
In FilenameFilterIterator.php line 43:
                                                   
  [Error]                                          
  Class "Symfony\Component\Finder\Glob" not found  
                                                   

Exception trace:
  at /app/vendor/symfony/finder/Iterator/FilenameFilterIterator.php:43
```

Bit I didn't have time to dig into this, could also be some edge cases from the existing project-template (phd5).

**Reply by terabytesoftw** [reply]
Could you tell me the versions you use to reproduce and fix it, composer show?

Symfony console.
npm,
yarn,
composer.

**Reply by schmunk42** [reply]
> Could you tell me the versions you use to reproduce and fix it, composer show?

I'll report directly to your repo if I find time to try this in-depth.

**Comment by mtangoo** [comment]
>If it is correct, if we move the assets to separate packages, it will break everything, because the dependencies must be added to the extensions.

But this is a composer one line. So it is not a big thing to do. Or do I miss something?

**Comment by terabytesoftw** [comment]
> > btw in Yii3 there is no download manager for the assets, you use whatever you want, from asset-packagist.org, foxy, or simply npm.
> 
> How do you use `asset-packagist.org` with Yii3, could you point me to the docs?

Very simple: https://github.com/yiisoft/yii-bootstrap5?tab=readme-ov-file#install-assets

**Comment by razvanphp** [comment]
This a far more extreme suggestion than what I had in mind with #20162

Anyway, I think the pain-points should be resolved, rather than create new breaking changes just to be on par with what's new on FE world today. For example, the bundling and minifier with java and external shell commands is way outdated, we should use something more established in the industry by now like parcel (with lighningcss).

Should not be so complicated for anybody that can do composer install to also run a few npm commands and run parcel itself. We should only make sure we have a valid `package.json` in core that can be used to install all needed deps (even with foxy).

I really love PHP, but I think we went too far with the whole asset management in Yii to avoid some simple, established ways to install JS packages. Today, many tools can install those deps besides npm, one can even use [bun](https://bun.sh/) or deno for example.

**Reply by schmunk42** [reply]
If someone has the time, here would be an idea for a PoC:
- Fork (all) yiisoft/yii2-... repos with assets
  - Remove `bower-`, `npm-` entries from `composer.json` in forked repos.
- Fork `yii2-advanced-app` or `yii2-basic-app`
  - Use `min-stability: dev`
  - Remove asset-packagist
  - Add forked repos to `composer.json`
  - Run `composer update`

If this works, step 1 is done :)

In the forked repos add a `package.json` with the corresponding assets and adjust asset bundles to `node_modules`.
Install `foxy` and run another `composer update` or install your assets manually or any tool you want.

Things should be working now.

**Reply by razvanphp** [reply]
but there is one big blockere for this plan, somebody from the yii team needs to publish the yii-assets to NPM (`yii.activeForm.js`, `yii.captcha.js`, `yii.gridView.js`, `yii.js`, `yii.validation.js`)

**Reply by mtangoo** [reply]
This should not be hard. If we agree on the most pertinent things, it will be done

**Reply by mtangoo** [reply]
@schmunk42 I will try to find time and test that!

**Reply by terabytesoftw** [reply]
Hi, i have done some testing and found out that you can remove the assets from the `composer.json`, just add them in the `package.json`, and when you install you can just use `npm` or `php-forge/foxy`, just adding it is an `app-basic` or `app-advanced`, and what you need to do is map `bower-asset` and `npm-asset` to `node_module`, something quite simple 😃

**Comment by razvanphp** [comment]
How "others" are doing it, using modern FE tooling: https://laravel.com/docs/11.x/vite

**Reply by terabytesoftw** [reply]
It would be great to update Yii2, not only the PHP versions, but part of its architecture, that would be a big BC, and a lot of work.

**Reply by razvanphp** [reply]
indeed, but let's try to retrofit at least some modern tools for asset handling, keeping the BC, I think this is doable....

**Comment by ItsReddi** [comment]
From my view, that should definitely take place in 2.2 since the current ecosystem is no longer reliable and maintained.
https://github.com/yiisoft/yii2/issues/20067

**Reply by schmunk42** [reply]
I can't find my longer post about this topic atm, but in a nutshell: **npm with composer is also not a viable solution.**

We need to remove assets from `composer.json` completely in 2.2 and use native `npm/yarn/...` oder `foxy`

**Reply by mtangoo** [reply]
Then load node_modules automatically? Is that what you are suggesting?
Making it as alias `@npm` should not be hard. Then use the alias to load assets!

**Reply by schmunk42** [reply]
Here's a PoC idea: https://github.com/yiisoft/yii2/discussions/20164#discussioncomment-9579392

I just want to stress, that using **`npm` with composer** is another dead-end. There are a huge amount of packages which use a different version of the same lib - which is not possible with composer, but no problem with native npm.

All asset packages need to be moved from `composer.json` to `packages.json`

**Reply by mtangoo** [reply]
I see. That will be a way cleaner and will give developer flexibility on tooling. I don't see why it shouldn't work

**Reply by schmunk42** [reply]
> I don't see why it shouldn't work

What do you mean specifically (removing asset packages or using npm packages with composer)?

**Reply by mtangoo** [reply]
Removing packages from composer and install them independently or using some tools line foxy thing.
Should work fine!

**Comment by diego-mathis** [comment]
Using assets with Yii2 should work without extras. In this sense the Yii Javascript libraries need to be released into the JS Oecosystem.
I had a look at foxy, it seems great - as it uses the specified underlying package manager.
But in general I believe, foxy should be sugar - making your life easier, as in installing nested dependencies.
However, not every PHP package is built around foxy, so there will be exceptions - where u need to add packages to packages.json despite using foxy.
So in my opinion, everything should be possible without foxy. 
Also if  the framework includes foxy, and foxy allows to use bun, npm .. etc...  wouldnt that require a test for all those variants?

**Reply by terabytesoftw** [reply]
As i explained in the PR, the framework in its `2.2` branch is free of any asset installer. It has its `package.json`, and that is enough to use any available package manager, be it `npm`, `bun`, `pnpm`, `yarn`, etc. In the PR i use `Foxy` because we already used it in the `2.2` branch, but unlike the `2.0` branch where `asset-packagist` is used, it is only a `dev` dependency, so the choice is up to the developer, you are free to choose. On the other hand, `Foxy` supports all existing asset managers: https://github.com/php-forge/foxy/tree/main/tests/Asset

btw this approach solves this master branch problem:

https://github.com/yiisoft/yii2/issues/20265#issuecomment-3347246850

**Comment by mtangoo** [comment]
Well, if you remove something like asset-packagist which wasn't developed by YiiSoft but a third party (an excellent contribution to the community) then the replacement should just work like the former.

For seasoned dev it is not an issue, but for beginners, it is. That's the gist of the discussion and the reason to put this package so that it will still auto install.

It is trading a third-party package to another. But for better results, if you wish.

So if not this, we still need to come with a solution to solve this problem.

**Comment by terabytesoftw** [comment]
done in 22 branch.
